operator/internal/certrotation/cert_creator.go

```diff
@@ -0,0 +1,178 @@
+package certrotation
+
+import (
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/openshift/library-go/pkg/certs"
+	"github.com/openshift/library-go/pkg/crypto"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/authentication/user"
+)
+
+var errMissingHostnames = errors.New("no hostnames set")
+
+// TargetCertCreator defines the interface of cert creating facilities.
+type TargetCertCreator interface {
+	// NewCertificate creates a new key-cert pair with the given signer.
+	NewCertificate(signer *crypto.CA, validity time.Duration) (*crypto.TLSCertificateConfig, error)
+	// NeedNewTargetCertKeyPair decides whether a new cert-key pair is needed. It returns a non-empty reason if it is the case.
+	NeedNewTargetCertKeyPair(currentSecretAnnotations map[string]string, signer *crypto.CA, caBundleCerts []*x509.Certificate, refresh time.Duration, refreshOnlyWhenExpired bool) string
+	// SetAnnotations gives an option to override or set additional annotations
+	SetAnnotations(cert *crypto.TLSCertificateConfig, annotations map[string]string) map[string]string
+}
+
+type clientCertCreator struct {
```

Comment:
Does it make sense to only have `servingCertCreator`, even for the gateway certificate? Technically the gateway only needs a client certificate with no hostname, but we could remove some of the duplication that way and make the code simpler.

As a side-note question: Is there anything stopping us from using the `ServingRotation` from `openshift/library-go`?

Reply:
The issue with using `openshift/library-go` for `ServingRotation` is that they hardcode the annotations keys: https://github.com/openshift/library-go/blob/master/pkg/operator/certrotation/client_cert_rotation_controller.go#L17-L28

This operator is not an openshift only operator and I would like to keep confusion out as much as possible.